File: apps/mobile-app/utils/db/mappers/FieldMapper.ts
import type { ItemField, FieldType } from '@/utils/dist/core/models/vault';
import { FieldTypes, getSystemField } from '@/utils/dist/core/models/vault';

/**
 * Raw field row from database query.
 */
export type FieldRow = {
  ItemId: string;
  FieldKey: string | null;
  FieldDefinitionId: string | null;
  CustomLabel: string | null;
  CustomFieldType: string | null;
  CustomIsHidden: number | null;
  CustomEnableHistory: number | null;
  Value: string;
  DisplayOrder: number;
};

/**
 * Intermediate field representation before grouping.
 */
export type ProcessedField = {
  ItemId: string;
  FieldKey: string;
  Label: string;
  FieldType: string;
  IsHidden: number;
  Value: string;
  DisplayOrder: number;
  IsCustomField: boolean;
  EnableHistory: boolean;
};

/**
 * Mapper class for processing database field rows into ItemField objects.
 * Handles both system fields (with FieldKey) and custom fields (with FieldDefinitionId).
 */
export class FieldMapper {
  /**
   * Process raw field rows from database into a map of ItemId -> ItemField[].
   * Handles system vs custom fields and multi-value field grouping.
   * @param rows - Raw field rows from database
   * @returns Map of ItemId to array of ItemField objects
   */
  public static processFieldRows(rows: FieldRow[]): Map<string, ItemField[]> {
    // First, convert rows to processed fields with proper metadata
    const processedFields = rows.map(row => this.processFieldRow(row));

    // Group by ItemId and FieldKey (to handle multi-value fields)
    const fieldsByItem = new Map<string, ItemField[]>();
    const fieldValuesByKey = new Map<string, string[]>();

    for (const field of processedFields) {
      const key = `${field.ItemId}_${field.FieldKey}`;

      // Accumulate values for the same field
      if (!fieldValuesByKey.has(key)) {
        fieldValuesByKey.set(key, []);
      }
      fieldValuesByKey.get(key)!.push(field.Value);

      // Create ItemField entry only once per unique FieldKey per item
      if (!fieldsByItem.has(field.ItemId)) {
        fieldsByItem.set(field.ItemId, []);
      }

      const itemFields = fieldsByItem.get(field.ItemId)!;
      const existingField = itemFields.find(f => f.FieldKey === field.FieldKey);

      if (!existingField) {
        itemFields.push({
          FieldKey: field.FieldKey,
          Label: field.Label,
          FieldType: field.FieldType as FieldType,
          Value: '', // Will be set below
          IsHidden: field.IsHidden === 1,
          DisplayOrder: field.DisplayOrder,
          IsCustomField: field.IsCustomField,
          EnableHistory: field.EnableHistory
        });
      }
    }

    // Set Values (single value or array for multi-value fields)
    for (const [itemId, fields] of fieldsByItem) {
      for (const field of fields) {
        const key = `${itemId}_${field.FieldKey}`;
        const values = fieldValuesByKey.get(key) || [];

        if (values.length === 1) {
          field.Value = values[0];
        } else {
          field.Value = values;
        }
      }
    }

    return fieldsByItem;
  }

  /**
   * Process a single field row to extract proper metadata.
   * System fields use FieldKey and get metadata from SystemFieldRegistry.
   * Custom fields use FieldDefinitionId and get metadata from the row.
   * @param row - Raw field row
   * @returns Processed field with proper metadata
   */
  private static processFieldRow(row: FieldRow): ProcessedField {
    if (row.FieldKey) {
      // System field: has FieldKey, get metadata from SystemFieldRegistry
      const systemField = getSystemField(row.FieldKey);
      return {
        ItemId: row.ItemId,
        FieldKey: row.FieldKey,
        Label: row.FieldKey, // Use FieldKey as label; UI layer translates via fieldLabels.*
        FieldType: systemField?.FieldType || FieldTypes.Text,
        IsHidden: systemField?.IsHidden ? 1 : 0,
        Value: row.Value,
        DisplayOrder: row.DisplayOrder,
        IsCustomField: false,
        EnableHistory: systemField?.EnableHistory ?? false
      };
    } else {
      // Custom field: has FieldDefinitionId, get metadata from FieldDefinitions
      return {
        ItemId: row.ItemId,
        FieldKey: row.FieldDefinitionId || '', // Use FieldDefinitionId (UUID) as the key for custom fields
        Label: row.CustomLabel || '',
        FieldType: row.CustomFieldType || FieldTypes.Text,
        IsHidden: row.CustomIsHidden || 0,
        Value: row.Value,
        DisplayOrder: row.DisplayOrder,
        IsCustomField: true,
        EnableHistory: row.CustomEnableHistory === 1
      };
    }
  }

  /**
   * Process field rows for a single item (without ItemId in result).
   * Used when fetching a single item by ID.
   * @param rows - Raw field rows for a single item
   * @returns Array of ItemField objects
   */
  public static processFieldRowsForSingleItem(rows: Omit<FieldRow, 'ItemId'>[]): ItemField[] {
    const fieldValuesByKey = new Map<string, string[]>();
    const uniqueFields = new Map<string, {
      FieldKey: string;
      Label: string;
      FieldType: string;
      IsHidden: number;
      DisplayOrder: number;
      IsCustomField: boolean;
      EnableHistory: boolean;
    }>();

    for (const row of rows) {
      const fieldKey = row.FieldKey || row.FieldDefinitionId || '';

      // Accumulate values
      if (!fieldValuesByKey.has(fieldKey)) {
        fieldValuesByKey.set(fieldKey, []);
      }
      fieldValuesByKey.get(fieldKey)!.push(row.Value);

      // Store field metadata (only once per FieldKey)
      if (!uniqueFields.has(fieldKey)) {
        if (row.FieldKey) {
          // System field
          const systemField = getSystemField(row.FieldKey);
          uniqueFields.set(fieldKey, {
            FieldKey: row.FieldKey,
            Label: row.FieldKey, // Use FieldKey as label; UI layer translates via fieldLabels.*
            FieldType: systemField?.FieldType || FieldTypes.Text,
            IsHidden: systemField?.IsHidden ? 1 : 0,
            DisplayOrder: row.DisplayOrder,
            IsCustomField: false,
            EnableHistory: systemField?.EnableHistory ?? false
          });
        } else {
          // Custom field
          uniqueFields.set(fieldKey, {
            FieldKey: fieldKey,
            Label: row.CustomLabel || '',
            FieldType: row.CustomFieldType || FieldTypes.Text,
            IsHidden: row.CustomIsHidden || 0,
            DisplayOrder: row.DisplayOrder,
            IsCustomField: true,
            EnableHistory: row.CustomEnableHistory === 1
          });
        }
      }
    }

    // Build fields array with proper single/multi values
    return Array.from(uniqueFields.entries()).map(([fieldKey, metadata]) => {
      const values = fieldValuesByKey.get(fieldKey) || [];
      return {
        ...metadata,
        FieldType: metadata.FieldType as FieldType,
        Value: values.length === 1 ? values[0] : values,
        IsHidden: metadata.IsHidden === 1,
        IsCustomField: metadata.IsCustomField,
        EnableHistory: metadata.EnableHistory
      };
    });
  }
}

File: apps/mobile-app/ios/VaultStoreKit/Database/Mappers/FieldMapper.swift
import Foundation
import VaultModels

/// Raw field row from database query.
public struct FieldRow {
    public let itemId: String
    public let fieldKey: String?
    public let fieldDefinitionId: String?
    public let customLabel: String?
    public let customFieldType: String?
    public let customIsHidden: Int64?
    public let customEnableHistory: Int64?
    public let value: String
    public let displayOrder: Int

    public init(
        itemId: String,
        fieldKey: String?,
        fieldDefinitionId: String?,
        customLabel: String?,
        customFieldType: String?,
        customIsHidden: Int64?,
        customEnableHistory: Int64?,
        value: String,
        displayOrder: Int
    ) {
        self.itemId = itemId
        self.fieldKey = fieldKey
        self.fieldDefinitionId = fieldDefinitionId
        self.customLabel = customLabel
        self.customFieldType = customFieldType
        self.customIsHidden = customIsHidden
        self.customEnableHistory = customEnableHistory
        self.value = value
        self.displayOrder = displayOrder
    }

    /// Initialize from a database row dictionary.
    public init?(from row: [String: Any]) {
        guard let itemId = row["ItemId"] as? String else { return nil }

        self.itemId = itemId
        self.fieldKey = row["FieldKey"] as? String
        self.fieldDefinitionId = row["FieldDefinitionId"] as? String
        self.customLabel = row["CustomLabel"] as? String
        self.customFieldType = row["CustomFieldType"] as? String
        self.customIsHidden = row["CustomIsHidden"] as? Int64
        self.customEnableHistory = row["CustomEnableHistory"] as? Int64
        self.value = row["Value"] as? String ?? ""
        self.displayOrder = Int(row["DisplayOrder"] as? Int64 ?? 0)
    }
}

/// Raw field row for single item queries (without ItemId).
public struct SingleItemFieldRow {
    public let fieldKey: String?
    public let fieldDefinitionId: String?
    public let customLabel: String?
    public let customFieldType: String?
    public let customIsHidden: Int64?
    public let customEnableHistory: Int64?
    public let value: String
    public let displayOrder: Int

    /// Initialize from a database row dictionary.
    public init?(from row: [String: Any]) {
        self.fieldKey = row["FieldKey"] as? String
        self.fieldDefinitionId = row["FieldDefinitionId"] as? String
        self.customLabel = row["CustomLabel"] as? String
        self.customFieldType = row["CustomFieldType"] as? String
        self.customIsHidden = row["CustomIsHidden"] as? Int64
        self.customEnableHistory = row["CustomEnableHistory"] as? Int64
        self.value = row["Value"] as? String ?? ""
        self.displayOrder = Int(row["DisplayOrder"] as? Int64 ?? 0)
    }
}

/// Intermediate field representation before grouping.
public struct ProcessedField {
    public let itemId: String
    public let fieldKey: String
    public let label: String
    public let fieldType: String
    public let isHidden: Bool
    public let value: String
    public let displayOrder: Int
    public let isCustomField: Bool
    public let enableHistory: Bool
}

/// Mapper class for processing database field rows into ItemField objects.
/// Handles both system fields (with FieldKey) and custom fields (with FieldDefinitionId).
public struct FieldMapper {
    /// Process raw field rows from database into a map of ItemId -> [ItemField].
    /// Handles system vs custom fields. Multi-value fields (like URLs) create separate ItemField entries
    /// with the same fieldKey but different values, allowing Item.getFieldValues() to return all values.
    /// - Parameter rows: Raw field rows from database
    /// - Returns: Dictionary of ItemId to array of ItemField objects
    public static func processFieldRows(_ rows: [FieldRow]) -> [String: [ItemField]] {
        // First, convert rows to processed fields with proper metadata
        let processedFields = rows.map { processFieldRow($0) }

        // Group fields by ItemId, keeping separate entries for multi-value fields
        var fieldsByItem: [String: [ItemField]] = [:]

        for field in processedFields {
            if fieldsByItem[field.itemId] == nil {
                fieldsByItem[field.itemId] = []
            }

            // Create an ItemField for each row (including duplicates for multi-value fields)
            let itemField = ItemField(
                fieldKey: field.fieldKey,
                label: field.label,
                fieldType: field.fieldType,
                value: field.value,
                isHidden: field.isHidden,
                displayOrder: field.displayOrder,
                isCustomField: field.isCustomField,
                enableHistory: field.enableHistory
            )
            fieldsByItem[field.itemId]!.append(itemField)
        }

        return fieldsByItem
    }

    /// Process a single field row to extract proper metadata.
    /// System fields use FieldKey and get metadata from SystemFieldRegistry.
    /// Custom fields use FieldDefinitionId and get metadata from the row.
    /// - Parameter row: Raw field row
    /// - Returns: Processed field with proper metadata
    private static func processFieldRow(_ row: FieldRow) -> ProcessedField {
        if let fieldKey = row.fieldKey, !fieldKey.isEmpty {
            // System field: has FieldKey, get metadata from field metadata resolver
            let metadata = resolveFieldMetadata(
                fieldKey: fieldKey,
                customLabel: nil,
                customFieldType: nil,
                customIsHidden: false,
                customEnableHistory: false,
                isCustomField: false
            )
            return ProcessedField(
                itemId: row.itemId,
                fieldKey: fieldKey,
                label: metadata.label,
                fieldType: metadata.fieldType,
                isHidden: metadata.isHidden,
                value: row.value,
                displayOrder: row.displayOrder,
                isCustomField: false,
                enableHistory: metadata.enableHistory
            )
        } else {
            // Custom field: has FieldDefinitionId, get metadata from FieldDefinitions
            let fieldKey = row.fieldDefinitionId ?? ""
            return ProcessedField(
                itemId: row.itemId,
                fieldKey: fieldKey,
                label: row.customLabel ?? "",
                fieldType: row.customFieldType ?? FieldType.text,
                isHidden: row.customIsHidden == 1,
                value: row.value,
                displayOrder: row.displayOrder,
                isCustomField: true,
                enableHistory: row.customEnableHistory == 1
            )
        }
    }

    /// Process field rows for a single item (without ItemId in result).
    /// Used when fetching a single item by ID.
    /// Multi-value fields (like URLs) create separate ItemField entries with the same fieldKey.
    /// - Parameter rows: Raw field rows for a single item
    /// - Returns: Array of ItemField objects
    public static func processFieldRowsForSingleItem(_ rows: [SingleItemFieldRow]) -> [ItemField] {
        // Create an ItemField for each row (including duplicates for multi-value fields)
        return rows.map { row in
            let fieldKey = row.fieldKey ?? row.fieldDefinitionId ?? ""
            let isCustomField = row.fieldKey == nil || row.fieldKey!.isEmpty

            if !isCustomField, let rowFieldKey = row.fieldKey {
                // System field
                let metadata = resolveFieldMetadata(
                    fieldKey: rowFieldKey,
                    customLabel: nil,
                    customFieldType: nil,
                    customIsHidden: false,
                    customEnableHistory: false,
                    isCustomField: false
                )
                return ItemField(
                    fieldKey: rowFieldKey,
                    label: metadata.label,
                    fieldType: metadata.fieldType,
                    value: row.value,
                    isHidden: metadata.isHidden,
                    displayOrder: row.displayOrder,
                    isCustomField: false,
                    enableHistory: metadata.enableHistory
                )
            } else {
                // Custom field
                return ItemField(
                    fieldKey: fieldKey,
                    label: row.customLabel ?? "",
                    fieldType: row.customFieldType ?? FieldType.text,
                    value: row.value,
                    isHidden: row.customIsHidden == 1,
                    displayOrder: row.displayOrder,
                    isCustomField: true,
                    enableHistory: row.customEnableHistory == 1
                )
            }
        }.sorted { $0.displayOrder < $1.displayOrder }
    }

    // MARK: - Field Metadata Resolution

    /// Helper struct to hold resolved field metadata.
    private struct FieldMetadata {
        let label: String
        let fieldType: String
        let isHidden: Bool
        let enableHistory: Bool
    }

    /// Resolve field metadata for system fields and custom fields.
    private static func resolveFieldMetadata(
        fieldKey: String,
        customLabel: String?,
        customFieldType: String?,
        customIsHidden: Bool,
        customEnableHistory: Bool,
        isCustomField: Bool
    ) -> FieldMetadata {
        if isCustomField {
            return FieldMetadata(
                label: customLabel ?? fieldKey,
                fieldType: customFieldType ?? FieldType.text,
                isHidden: customIsHidden,
                enableHistory: customEnableHistory
            )
        }

        // System field metadata based on FieldKey constants
        switch fieldKey {
        case FieldKey.loginUsername:
            return FieldMetadata(label: "Username", fieldType: FieldType.text, isHidden: false, enableHistory: false)
        case FieldKey.loginPassword:
            return FieldMetadata(label: "Password", fieldType: FieldType.password, isHidden: true, enableHistory: true)
        case FieldKey.loginEmail:
            return FieldMetadata(label: "Email", fieldType: FieldType.email, isHidden: false, enableHistory: false)
        case FieldKey.loginUrl:
            return FieldMetadata(label: "URL", fieldType: FieldType.uRL, isHidden: false, enableHistory: false)
        case FieldKey.cardNumber:
            return FieldMetadata(label: "Card Number", fieldType: FieldType.text, isHidden: true, enableHistory: false)
        case FieldKey.cardCardholderName:
            return FieldMetadata(label: "Cardholder Name", fieldType: FieldType.text, isHidden: false, enableHistory: false)
        case FieldKey.cardExpiryMonth:
            return FieldMetadata(label: "Expiry Month", fieldType: FieldType.text, isHidden: false, enableHistory: false)
        case FieldKey.cardExpiryYear:
            return FieldMetadata(label: "Expiry Year", fieldType: FieldType.text, isHidden: false, enableHistory: false)
        case FieldKey.cardCvv:
            return FieldMetadata(label: "CVV", fieldType: FieldType.password, isHidden: true, enableHistory: false)
        case FieldKey.cardPin:
            return FieldMetadata(label: "PIN", fieldType: FieldType.password, isHidden: true, enableHistory: false)
        case FieldKey.aliasFirstName:
            return FieldMetadata(label: "First Name", fieldType: FieldType.text, isHidden: false, enableHistory: false)
        case FieldKey.aliasLastName:
            return FieldMetadata(label: "Last Name", fieldType: FieldType.text, isHidden: false, enableHistory: false)
        case FieldKey.aliasGender:
            return FieldMetadata(label: "Gender", fieldType: FieldType.text, isHidden: false, enableHistory: false)
        case FieldKey.aliasBirthdate:
            return FieldMetadata(label: "Birth Date", fieldType: FieldType.date, isHidden: false, enableHistory: false)
        case FieldKey.notesContent:
            return FieldMetadata(label: "Notes", fieldType: FieldType.textArea, isHidden: false, enableHistory: false)
        default:
            // Unknown system field - use field key as label
            return FieldMetadata(label: fieldKey, fieldType: FieldType.text, isHidden: false, enableHistory: false)
        }
    }
}
